image/manifest.go

```diff
@@ -59,54 +300,60 @@ func ReadSeedManifest(manifestFile string) (map[string]snap.Revision, error) {
 		}
 
 		tokens := strings.Fields(line)
-		// Expect exactly two tokens
-		if len(tokens) != 2 {
-			return nil, fmt.Errorf("line is illegally formatted: %q", line)
-		}
 
-		snapName := tokens[0]
-		revString := tokens[1]
-		if err := snap.ValidateName(snapName); err != nil {
-			return nil, err
-		}
-
-		rev, err := snap.ParseRevision(revString)
-		if err != nil {
-			return nil, err
-		}
-
-		// Values that are higher than 0 indicate the revision comes from the store, and values
-		// lower than 0 indicate the snap was sourced locally. We allow both in the seed.manifest as
-		// long as the user can provide us with the correct snaps. The only number we won't accept is
-		// 0.
-		if rev.Unset() {
-			return nil, fmt.Errorf("cannot use revision %d for snap %q: revision must not be 0", rev, snapName)
+		if len(tokens) == 1 && strings.Contains(tokens[0], "/") {
+			// Pinned validation-set: <account-id>/<name>=<sequence>
+			if err := parsePinnedValidationSet(sm, tokens[0]); err != nil {
+				return nil, err
+			}
+		} else if len(tokens) == 2 {
+			if strings.Contains(tokens[0], "/") {
+				// Unpinned validation-set: <account-id>/<name> <sequence>
+				if err := parseUnpinnedValidationSet(sm, tokens[0], tokens[1]); err != nil {
+					return nil, err
+				}
+			} else {
+				// Snap revision: <snap> <revision>
+				if err := parseSnapRevision(sm, tokens[0], tokens[1]); err != nil {
+					return nil, err
+				}
+			}
+		} else {
+			return nil, fmt.Errorf("line is illegally formatted: %q", line)
 		}
-		revisions[snapName] = rev
 	}
-	return revisions, nil
+	return sm, nil
 }
 
-// WriteSeedManifest generates the seed.manifest contents from the provided map of
-// snaps and their revisions, and stores them in the given file path
-func WriteSeedManifest(filePath string, revisions map[string]snap.Revision) error {
-	if len(revisions) == 0 {
+// Write generates the seed.manifest contents from the provided map of
+// snaps and their revisions, and stores them in the given file path.
+func (sm *SeedManifest) Write(filePath string) error {
+	if len(sm.used) == 0 {
 		return nil
 	}
 
-	keys := make([]string, 0, len(revisions))
-	for k := range revisions {
-		keys = append(keys, k)
+	vsKeys := make([]string, 0, len(sm.used))
+	for k, s := range sm.used {
+		if _, ok := s.(*SeedManifestValidationSet); ok {
+			vsKeys = append(vsKeys, k)
+		}
 	}
-	sort.Strings(keys)
+	sort.Strings(vsKeys)
 
-	buf := bytes.NewBuffer(nil)
-	for _, key := range keys {
-		rev := revisions[key]
-		if rev.Unset() {
-			return fmt.Errorf("revision must not be 0 for snap %q", key)
+	revisionKeys := make([]string, 0, len(sm.used))
+	for k, s := range sm.used {
+		if _, ok := s.(*SeedManifestSnapRevision); ok {
```

Comment:
same for used, also I'm not sure when we need to deal with this but we probably shouldn't register revisions for snap that are covered already by a validation set constraint that includes a revision